wire-ios-sync-engine/Source/Services/SupportedProtocolsService.swift

```diff
@@ -0,0 +1,184 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+protocol SupportedProtocolsServiceInterface {
+
+    func updateSupportedProtocols()
+
+}
+
+final class SupportedProtocolsService: SupportedProtocolsServiceInterface {
+
+    // MARK: - Properties
+
+    private let featureRepository: FeatureRepositoryInterface
+    private let userRepository: UserRepositoryInterface
+
+    // MARK: - Life cycle
+
+    public convenience init(context: NSManagedObjectContext) {
+        self.init(
+            featureRepository: FeatureRepository(context: context),
+            userRepository: UserRepository(context: context)
+        )
+    }
+
+    init(
+        featureRepository: FeatureRepositoryInterface,
+        userRepository: UserRepositoryInterface
+    ) {
+        self.featureRepository = featureRepository
+        self.userRepository = userRepository
+    }
+
+    // MARK: - Methods
+
+    func updateSupportedProtocols() {
+        let selfUser = userRepository.selfUser()
+        selfUser.supportedProtocols = calculateSupportedProtocols()
+    }
+
+    func calculateSupportedProtocols() -> Set<MessageProtocol> {
+        let remoteProtocols = remotelySupportedProtocols()
+        let migrationState = currentMigrationState()
+        let allClientsMLSReady = allSelfUserClientsAreActiveMLSClients()
+
+        var result = Set<MessageProtocol>()
+
+        // All clients are proteus ready so we support it if the backend does.
+        if remoteProtocols.contains(.proteus) {
+            result.insert(.proteus)
+        }
+
+        // All clients are mls ready so we support it if the backend does.
+        if remoteProtocols.contains(.mls) && allClientsMLSReady {
+            result.insert(.mls)
+        }
+
+        // Proteus is still supported if migration is pending or still ongoing.
+        if migrationState.isOne(of: .notStarted, .ongoing) && allClientsMLSReady {
+            result.insert(.proteus)
+        }
+
+        // MLS migration is complete.
+        if remoteProtocols.contains(.mls) && migrationState == .finalised {
+            result.insert(.mls)
+        }
+
+        // MLS is forced.
+        if remoteProtocols == [.mls] && migrationState.isOne(of: .disabled, .finalised) {
+            result = [.mls]
+        }
+
+        // Even if proteus isn't supported, migration is pending or still ongoing.
+        if remoteProtocols == [.mls] && !allClientsMLSReady && migrationState.isOne(of: .notStarted, .ongoing) {
+            result = [.proteus]
+        }
```

Comment:
how come we force proteus if it's not supported. It's not mixed?

Reply:
All clients support proteus, only some support mls. The team admin may say they only want to use mls and not proteus. In this particular case, the team admin has opted for a migration (since the migration state is not started or ongoing). 

The self client here only supports proteus because not all of their clients are mls ready. If they upgrade all their clients, then they will support switch support to mls only (in line 72).  If the migration has been finalized (i.e deadline to upgrade all your clients has passed), then the self client will support only mls even if all their clients haven't been upgraded (in line 87).